### PR TITLE
`fn ipred_z*_neon`: Cleanup and make safer

### PR DESCRIPTION
**Pull request by randomPoison**
* Part of #846.
* Part of #815.

### CODE REVIEW

File: src/ipred.rs [hunk 1]
```diff
@@ -1941,13 +1928,13 @@ mod neon {
...
                 get_filter_strength(width + height, angle - 180, is_sm)
             } else {
-                0 as c_int
+                0
             };
             if filter_strength != 0 {
-                flipped[0] = *topleft_in.offset(0);
+                flipped[0] = topleft_in[0];
                 bd_fn!(reverse::decl_fn, BD, ipred_reverse, neon).call::<BD>(
                     flipped.as_mut_ptr().offset(1),
```
Comment (kkysen) [inline src/ipred.rs:1936]:
```suggestion
                    flipped[1..].as_mut_ptr(),
```

File: src/ipred.rs [hunk 2]
```diff
@@ -2007,12 +1995,15 @@ mod neon {
         max_width: c_int,
         max_height: c_int,
         bitdepth_max: c_int,
-        _topleft_off: usize,
+        topleft_off: usize,
     ) {
+        // SAFETY: Reconstructed from args passed by `angular_ipred::Fn::call`.
```
Comment (kkysen) [inline src/ipred.rs:2000]: Can you also add the `# Safety` comment on `fn ipred_z_neon_erased` saying it must be called by ``[`angular_ipred::Fn::call`]``?

**Review by kkysen — CHANGES_REQUESTED** [review]
Can you also move the `.as_ptr()`/`.as_mut_ptr()`-type calls to the `fn call`s instead of repeating them at the call sites?

File: src/ipred.rs [hunk 3]
```diff
@@ -1985,6 +1984,7 @@ mod neon {
         };
     }
 
+    /// SAFETY: Must be called from [`angular_ipred::Fn::call`].
```
Comment (kkysen) [inline src/ipred.rs:1987]:
```suggestion
    /// # Safety
    ///
    /// Must be called from [`angular_ipred::Fn::call`].
```

`# Safety` is used for docs, vs. `// SAFETY` for comments for `unsafe {}` blocks.

**Review by kkysen — CHANGES_REQUESTED** [review]
Other new changes look good, but you can do this, too? https://github.com/memorysafety/rav1d/pull/1184#pullrequestreview-2106061876

**Comment by randomPoison** [comment]
> Can you also move the `.as_ptr()`/`.as_mut_ptr()`-type calls to the `fn call`s instead of repeating them at the call sites?

I'm not sure what you're asking for here. Are you referring to the places in the `ipred_z*_neon` fns where we pass a pointer to a buffer into the asm routines? So e.g. instead of:

```rust
let buf = [0; N];
some_asm_call(buf.as_ptr());
another_asm_call(buf.as_ptr());
```

we'd do:

```rust
let buf = [0; N];
let buf_ptr = buf.as_ptr();
some_asm_call(buf_ptr);
another_asm_call(buf_ptr);
```

Is that what you had in mind? That doesn't really seem any cleaner to me, it seems better to convert to a pointer at the call site. Am I misunderstanding what you're suggesting?

**Comment by kkysen** [comment]
> > Can you also move the `.as_ptr()`/`.as_mut_ptr()`-type calls to the `fn call`s instead of repeating them at the call sites?
> 
> I'm not sure what you're asking for here. Are you referring to the places in the `ipred_z*_neon` fns where we pass a pointer to a buffer into the asm routines? So e.g. instead of:
> 
> ```rust
> let buf = [0; N];
> some_asm_call(buf.as_ptr());
> another_asm_call(buf.as_ptr());
> ```
> 
> we'd do:
> 
> ```rust
> let buf = [0; N];
> let buf_ptr = buf.as_ptr();
> some_asm_call(buf_ptr);
> another_asm_call(buf_ptr);
> ```
> 
> Is that what you had in mind? That doesn't really seem any cleaner to me, it seems better to convert to a pointer at the call site. Am I misunderstanding what you're suggesting?

No, I meant, we're doing things like `.call::<BD>(buf[..n].as_ptr())` repeatedly instead of doing `.call::<BD(&buf[..n])` and doing the `let buf = buf.as_ptr()` inside of the `fn call`.  The approach we've been doing for most of the `fn call`s wrapping `fn` ptrs.

**Comment by randomPoison** [comment]
Ah, okay that makes sense. I've made that change.

File: src/ipred.rs [hunk 4]
```diff
@@ -1820,10 +1820,11 @@ mod neon {
                     topleft_in.as_ptr(),
                     height,
                 );
+                let (front, back) = buf.split_at_mut(1 + left_offset);
```
Comment (kkysen) [inline src/ipred.rs:1810]: Maybe `out` and `in_0` are better names, corresponding to the arg names.

File: src/ipred.rs [hunk 5]
```diff
@@ -1786,10 +1786,11 @@ mod neon {
                 topleft_in,
                 height,
             );
+            let (front, back) = buf.split_at_mut(left_offset);
```
Comment (kkysen) [inline src/ipred.rs:1789]: Same here.

**Review by kkysen — CHANGES_REQUESTED** [review]
One little thing left: for the places where you did `let (front, back) = _.split_at_mut(...);`, can you name them `in`, `out` or `src`, `dst` instead, for clarity, and matching the arg names and var names used elsewhere?

**Comment by randomPoison** [comment]
Done, I went with `src` and `dst`.

**Review by kkysen — APPROVED** [review]

**Pull request merged**